src/hotspot/share/gc/shared/partialArrayState.hpp

```diff
@@ -91,47 +95,135 @@ class PartialArrayState {
 
   NONCOPYABLE(PartialArrayStateAllocator);
 
   // Create a new state, obtaining the memory for it from the free-list or
-  // arena associated with worker_id.
-  PartialArrayState* allocate(uint worker_id,
-                              oop src, oop dst,
+  // from the associated manager.
+  PartialArrayState* allocate(oop src, oop dst,
                               size_t index, size_t length,
                               size_t initial_refcount);
 
   // Decrement the state's refcount.  If the new refcount is zero, add the
   // state to the free-list associated with worker_id.  The state must have
   // been allocated by this allocator, but that allocation doesn't need to
   // have been associated with worker_id.
-  void release(uint worker_id, PartialArrayState* state);
+  void release(PartialArrayState* state);
+};
+
+// This class provides memory management for PartialArrayStates.
+//
+// States are allocated using an allocator object. Those allocators in turn
+// may request memory for a state from their associated manager. The manager
+// is responsible for obtaining and releasing memory used for states by the
+// associated allocators.
+//
+// A state may be allocated by one allocator, but end up on the free-list of a
+// different allocator.  This can happen because a task referring to the state
+// may be stolen from the queue where it was initially added.  This is permitted
+// because a state's memory won't be reclaimed until all of the allocators
+// associated with the manager that is ultimately providing the memory have
+// been deleted and the manager is reset.
+//
+// A manager is used in two distinct and non-overlapping phases.
+//
+// - allocating: This is the initial phase.  During this phase, new allocators
+// may be created, and allocators may request memory from the manager.
+//
+// - releasing: When an allocator is destroyed the manager transitions to this
+// phase.  It remains in this phase until all extent allocators associated with
+// this manager have been destroyed.  During this phase, new allocators man not
+// be created, nor may extent allocators request memory from this manager.
+//
+// Once all the associated allocators have been destroyed the releasing phase
+// ends and the manager may be reset or deleted.  Resetting transitions back
+// to the allocating phase.
+class PartialArrayStateManager : public CHeapObj<mtGC> {
+  friend class PartialArrayStateAllocator;
+
+  // Use an arena for each allocator, for thread-safe concurrent allocation by
+  // different allocators.
+  Arena* _arenas;
+
+  // Limit on the number of allocators this manager supports.
+  uint _num_allocators;
+
+  // Integral type large enough to encode a pair of allocator counters as a
+  // single unit for atomic manipulation.
+  using CounterState = uintx;
+
+  // Encodes 2 values, in an atomic unit.
+  // - low half: allocators constructed
+  // - high half: allocators destructed (debug only)
+  volatile CounterState _counters;
```

Comment:
I feel the motivation for this encoding is missing from this PR. It's not immediately clear why these two counters need to be combined in this way. Kim outlined some rationale during our offline discussion, but for the benefit of other reviewers and future readers of this code, this rationale should be documented alongside the encoding. Having those arguments written down would help assess whether the additional complexity is justified.

Reply:
What about using a union/struct instead of all that manual masking and shifting?

I agree that encoding them in this way should give a rationale for that.

Reply:
I tried using some combination of union/struct a couple of different ways, but
the the result seemd significantly worse than what I published. But some
simplifications since those attempts, plus an idea for a different way of
structuring things, and I've now got something union/struct-based that seems
better.

It could be slightly further improved in syntax if I was willing to drop the
CounterState::_cf member name and make it an anonymous struct. That's a C11
feature that is provided as a C++ extension by all of our supported compilers.

You might not want to look at the difference between the two commits very
much, but instead look at the updated complete change.

Reply:
I also added expanded commentary for `PartialArrayStateManager::_counters`.  Maybe that will help?